### PR TITLE
MF-150: Add camera feature to attachments widget

### PR DESCRIPTION
**Pull request by ivange94**
Adds ability to capture attachments on the patient charts using the webcam.

### CODE REVIEW

File: src/widgets/attachments/attachments-overview.component.tsx [hunk 1]
```diff
@@ -124,6 +132,46 @@ export default function AttachmentsOverview() {
     }
   }
 
+  function openCamera(e: React.SyntheticEvent) {
+    setCameraIsOpen(true);
+  }
+
+  function handleTakePhoto(dataUri) {
```
Comment (FlorianRappl) [inline src/widgets/attachments/attachments-overview.component.tsx]: `dataUri: string`?

File: src/widgets/attachments/attachments-overview.component.tsx [hunk 2]
```diff
@@ -143,6 +191,24 @@ export default function AttachmentsOverview() {
             onChange={e => handleUpload(e, e.target.files)}
           />
         </form>
+        <div className={styles.cameraSection}>
+          <button className="cameraButton" onClick={openCamera}>
+            Camera
+          </button>
+          {cameraIsOpen && (
```
Comment (FlorianRappl) [inline src/widgets/attachments/attachments-overview.component.tsx]: I think this whole thing should be moved to a new component. Ideally, the whole state related to camera is in this component.

Something like `<Camera />`. As props we'd have something like `onNewImage`, which is when a camera component informs us that a new image has been successfully uploaded and should be added to the gallery.

Would that make sense?

File: src/widgets/attachments/attachments-overview.css [hunk 3]
```diff
@@ -7,3 +7,12 @@
   padding: 6px 12px;
   cursor: pointer;
 }
+
+.cameraSection {
+  padding: 10px;
+}
+
+.react-html5-camera-photo > img,
```
Comment (FlorianRappl) [inline src/widgets/attachments/attachments-overview.css]: Are these working? I thought the class names are prefixed and do not apply.

**Review by FlorianRappl — COMMENTED** [review]
Some minor comments.

File: src/widgets/attachments/attachments-overview.css (hunk 3 above)
Comment (ivange94) [reply src/widgets/attachments/attachments-overview.css]: Sorry those aren't working. That was why I created a global style and put them there. I must have forgot to delete here.

File: src/widgets/attachments/attachments-overview.component.tsx (hunk 2 above)
Comment (ivange94) [reply src/widgets/attachments/attachments-overview.component.tsx]: This makes sense. But do we have to name the component Camera? I mean I library already provides a component with that name. Will this act as a wrapper?

Comment (FlorianRappl) [reply src/widgets/attachments/attachments-overview.component.tsx]: Well, first of all its about the reusability discussed last time, but second it should bring the lib into the business domain of OpenMRS. As such the API communication for upload should be part of that (as described).

The name was freely chosen. If you'd see a better name (maybe `CameraUpload`?) feel free to choose one.

File: src/widgets/attachments/camera-upload.component.tsx [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+  function openCamera(e: React.SyntheticEvent) {
+    setCameraIsOpen(true);
+  }
+
+  function handleCloseCamera() {
+    setCameraIsOpen(false);
+  }
+
+  function handleTakePhoto(dataUri) {
```
Comment (FlorianRappl) [inline src/widgets/attachments/camera-upload.component.tsx]: `dataUri: string`

File: src/widgets/attachments/camera-upload.component.tsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  const [dataUri, setDataUri] = useState("");
+
+  const [
+    isLoadingPatient,
+    patient,
+    patientUuid,
+    patientErr
+  ] = useCurrentPatient();
+
+  function openCamera(e: React.SyntheticEvent) {
```
Comment (FlorianRappl) [inline src/widgets/attachments/camera-upload.component.tsx]: `e` is not used

File: src/widgets/attachments/camera-upload.component.tsx [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        props.onNewAttachment(att);
+        setDataUri("");
+      }
+    );
+  }
+
+  return (
+    <div className={styles.cameraSection}>
+      <button className="cameraButton" onClick={openCamera}>
+        Camera
```
Comment (FlorianRappl) [inline src/widgets/attachments/camera-upload.component.tsx]: Translatable string.

File: src/widgets/attachments/image-preview.component.tsx [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    e.preventDefault();
+    props.onCancelCapture();
+  }
+
+  function handleSubmit(e: React.SyntheticEvent) {
+    e.preventDefault();
+    e.stopPropagation();
+  }
+
+  function updateCaption(e: React.SyntheticEvent) {
```
Comment (FlorianRappl) [inline src/widgets/attachments/image-preview.component.tsx]: Prefer to use `React.ChangeEvent<HTMLInputElement>`. Then we also don't need to cast below.

File: src/widgets/attachments/image-preview.component.tsx [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    const target = e.target as HTMLInputElement;
+    setCaption(target.value);
+  }
+
+  return (
+    <form className={styles.overview} onSubmit={handleSubmit}>
+      <img src={props.dataUri} alt="Webcam preview" />
+      <input
+        type="text"
+        placeholder="Enter a caption for the image"
```
Comment (FlorianRappl) [inline src/widgets/attachments/image-preview.component.tsx]: Translatable string.
(same is true for the other strings here, e.g., `Webcam preview`, `Save`, `Cancel`.

File: src/widgets/attachments/attachments-overview.component.tsx [hunk 9]
```diff
@@ -124,6 +125,12 @@ export default function AttachmentsOverview() {
     }
   }
 
+  function handleNewAttachment(att) {
```
Comment (FlorianRappl) [inline src/widgets/attachments/attachments-overview.component.tsx]: What's the type of `att`?

**Review by FlorianRappl — COMMENTED** [review]
I think we are on a good way here!

**Review by FlorianRappl — APPROVED** [review]
LGTM! Great job Larry :beers:

**Comment by FlorianRappl** [comment]
It seems we got in some conflicts - can you try updating / resolving them? Thanks!

cc @ivange94

**Comment by ivange94** [comment]
> It seems we got in some conflicts - can you try updating / resolving them? Thanks!
> 
I just did. Should I squash the commits too?

**Comment by FlorianRappl** [comment]
Its all fine - great job again!

**Pull request merged**